### PR TITLE
Matchmaker

### PR DESCRIPTION
**Pull request by audebert**
TODO:

- [x] Add ansible setup
- [x] Export prometheus metrics

### CODE REVIEW

File: python-lib/prologin/matchmaker/matchmaker.py [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+        """Create new matches at predefined rate per second."""
+        while True:
+            if self.next_matches:
+                logging.info('%d matches in queue', len(self.next_matches))
+
+            new_matches = []
+            while (
+                len(new_matches)
+                < self.config["matchmaker"]["matches_per_second"]
+                and len(self.next_matches) != 0
```
Comment (Zopieux) [inline python-lib/prologin/matchmaker/matchmaker.py:288]: `and self.next_matches`?

File: python-lib/prologin/matchmaker/matchmaker.py [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+            ):
+                # Peek first
+                new_match = self.next_matches[0]
```
Comment (Zopieux) [inline python-lib/prologin/matchmaker/matchmaker.py:291]: `first_match, *self.next_matches = self.next_matches`

File: python-lib/prologin/matchmaker/matchmaker.py [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+    async def match_creator_task(self):
+        """Create new matches at predefined rate per second."""
+        while True:
+            if self.next_matches:
+                logging.info('%d matches in queue', len(self.next_matches))
+
+            new_matches = []
+            while (
+                len(new_matches)
+                < self.config["matchmaker"]["matches_per_second"]
```
Comment (Zopieux) [inline python-lib/prologin/matchmaker/matchmaker.py:287]: Looks like this is also missing a max size check, currently this can grow indefinitely (at a fixed rate).

File: python-lib/prologin/matchmaker/matchmaker.py [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+    Map,
+    Match,
+    MatchPriority,
+    Tournament,
+)
+
+
+@dataclasses.dataclass
+class MatchItem:
+    '''Class that represents a MatchMaker match.'''
```
[the remark this answers is not in the document]
Comment (audebert) [reply python-lib/prologin/matchmaker/matchmaker.py:44]: Done.

File: python-lib/prologin/matchmaker/matchmaker.py [hunk 5]
```diff
@@ -0,0 +1,308 @@
...
+        # TODO: use collections.dequeue
+        self.next_matches = []
+
+    def run(self):
+        logging.info(
+            "matchmaker listening on port %s",
+            self.config["matchmaker"]["port"],
+        )
+
+        self.run_task = asyncio.Task(self.run_task())
```
[the remark this answers is not in the document]
Comment (audebert) [reply python-lib/prologin/matchmaker/matchmaker.py:94]: Good catch. Renamed to start/start_task.

File: python-lib/prologin/matchmaker/matchmaker.py [hunk 6]
```diff
@@ -0,0 +1,308 @@
...
+            .values("max_id")
+        )
+
+        return Champion.objects.filter(pk__in=chs_ids)
+
+    async def setup(self):
+        self.author = django.contrib.auth.get_user_model().objects.get(
+            pk=self.config["matchmaker"]["author_id"]
+        )
+        logging.info("setup: match author: %s", self.author)
```
[the remark this answers is not in the document]
Comment (audebert) [reply python-lib/prologin/matchmaker/matchmaker.py:131]: Good point. Fixed.

File: python-lib/prologin/matchmaker/matchmaker.py [hunk 7]
```diff
@@ -0,0 +1,308 @@
...
+        self.add_matches(missing_matches)
+
+    def get_desired_matches(self, champions):
+        """Yields all expected matches for champions."""
+        for chs in itertools.product(
+            champions, repeat=django_settings.STECHEC_NPLAYERS
+        ):
+            # Don't fight against yourself
+            ch_ids = [c.id for c in chs]
+            if len(set(ch_ids)) != len(ch_ids):
```
[the remark this answers is not in the document]
Comment (audebert) [reply python-lib/prologin/matchmaker/matchmaker.py:194]: We don't just want a N-combination of the champions.

What the current algorithm produces is both a combinations then a permutation without duplicate entries, so that each champion runs against all other champions, in all possible orders. Using the full product seems to be the simplest and clearest way to achieve that.

File: python-lib/prologin/matchmaker/matchmaker.py [hunk 8]
```diff
@@ -0,0 +1,308 @@
...
+                self.matches_with_champion[champion].append(match)
+        # Commit by replacing old match list to avoid race conditions
+        self.next_matches = self.next_matches + matches
+
+    async def cancel_matches(self, old_champions, current_champions):
+        logging.info('Cancelling matches by champions %s', old_champions)
+
+        # Cancel scheduled matches
+        for old_champion in old_champions:
+            for match in self.matches_with_champion[old_champion]:
```
[the remark this answers is not in the document]
Comment (audebert) [reply python-lib/prologin/matchmaker/matchmaker.py:220]: Thanks!

File: python-lib/prologin/matchmaker/matchmaker.py (hunk 1 above)
Comment (audebert) [reply python-lib/prologin/matchmaker/matchmaker.py:288]: Thanks

File: python-lib/prologin/matchmaker/matchmaker.py (hunk 3 above)
Comment (audebert) [reply python-lib/prologin/matchmaker/matchmaker.py:287]: The max size of `new_matches` is `matches_per_second`. The virtual list that will grow indefinitely is the number of launched matches, which are scheduled just after this `while` loop.

We do need a global limit to how many matches MatchMaker can create, to avoid overloading masternode. I'll add that shortly.

File: python-lib/prologin/matchmaker/matchmaker.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (audebert) [reply python-lib/prologin/matchmaker/matchmaker.py]: Discussed offline → now using itertools.permutations().

File: python-lib/prologin/matchmaker/matchmaker.py (hunk 2 above)
Comment (audebert) [reply python-lib/prologin/matchmaker/matchmaker.py]: Now using deque().

File: python-lib/prologin/matchmaker/matchmaker.py [hunk 9]
```diff
@@ -0,0 +1,308 @@
...
+        cancelled_matches.update(status='cancelled')
+        for match in cancelled_matches.all():
+            match.remove(self.tournament)
+
+    async def schedule_new_matches(self, known_champions, new_champions):
+        logging.info('Scheduling matches for new champions %s', new_champions)
+
+        new_matches = []
+        # Iterate over new matches
+        for chs in itertools.product(
```
[the remark this answers is not in the document]
Comment (audebert) [reply python-lib/prologin/matchmaker/matchmaker.py]: Now using itertools.permutations().

File: python-lib/prologin/matchmaker/matchmaker.py [hunk 10]
```diff
@@ -131,7 +133,8 @@ def run(self, *args, **kwargs):
             self.config["matchmaker"]["port"],
         )
 
-        asyncio.create_task(self.run_task())
+        loop = asyncio.get_event_loop()
+        loop.create_task(self.run_task())
```
Comment (seirl) [inline python-lib/prologin/matchmaker/matchmaker.py:137]: Why is that needed? In modern asyncio you should never have to use loops manually when using the high level API.

Comment (audebert) [reply python-lib/prologin/matchmaker/matchmaker.py]: We cannot use `create_task()` here because there is no loop when this function runs and it raises `RuntimeError`.

In [other services](https://github.com/prologin/sadm/blob/276ed79b53d53cd0b486b302b942e0442ce57e2b/prologin/masternode/master.py#L59) we use `asyncio.Task()`, but that's discouraged too.

We cannot make `run()` a coro and use `asyncio.run(s.run())` either because that clashes with [aiohttp's model](https://github.com/aio-libs/aiohttp/blob/5f0a59fd38f048ee65b6199a26d2355075d0d196/aiohttp/web.py#L447), which expects to start the loop with `run_until_complete()`.

I'm now proposing an alternate solution, which uses aiohttp's [`on_startup` signal](https://docs.aiohttp.org/en/stable/web_advanced.html#aiohttp-web-signals) to start the service once the web app is started.

We may want to consider refactoring the other services with this pattern.

Comment (seirl) [reply python-lib/prologin/matchmaker/matchmaker.py]: Hmm, I don't think the other services would have the same issue, as the loop is already running when these tasks are created.

Comment (audebert) [reply python-lib/prologin/matchmaker/matchmaker.py]: Are you sure? In https://github.com/prologin/sadm/blob/master/prologin/masternode/master.py#L57-L66 the loop is started in `super().run()`.

Comment (seirl) [reply python-lib/prologin/matchmaker/matchmaker.py]: Indeed, I stand corrected!

**Comment by audebert** [comment]
Final call for reviews, let's merge this tomorrow.

File: debugconfig/matchmaker.yml [hunk 11]
```diff
@@ -0,0 +1,15 @@
+# MatchMaker server settings.
+matchmaker:
+  port: 8069 # RPC endpoint
+
+  shared_secret: "%%SECRET:cluster%%" # A shared secret used to avoid malicious
+                                      # impersonating requests.
```
Comment (Zopieux) [inline debugconfig/matchmaker.yml:6]: Nit: comment at the op, not inline on two lines.

File: prologin/matchmaker/__init__.py [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+new champions.
+
+MatchMaker configuration elements are:
+
+* **port** the TCP port to listen on.
+* **author_username** the username of the user to use to create matches
+* **tournament_name** the name of the tournament to create matches in
+* **include_staff** a boolean, whether staff champions are included in the
+  tournament
+* **deadline** a ISO 8601 date after which champions are not accepted
```
Comment (Zopieux) [inline prologin/matchmaker/__init__.py:14]: an

File: prologin/matchmaker/__init__.py [hunk 13]
```diff
@@ -0,0 +1,31 @@
...
+  (optional)
+* **match_repeat** the replication factor of created matches, 1 is the default,
+  2 means each unique match is created twice
```
Comment (Zopieux) [inline prologin/matchmaker/__init__.py:17]: Maybe clarify why this is useful, eg. something about randomness.

File: prologin/matchmaker/__main__.py [hunk 14]
```diff
@@ -0,0 +1,64 @@
...
+if __name__ == "__main__":
+    # Argument parsing
+    parser = optparse.OptionParser()
+    parser.add_option(
+        "-l",
+        "--local-logging",
+        action="store_true",
+        dest="local_logging",
+        default=False,
+        help="Activate logging to stdout.",
```
Comment (Zopieux) [inline prologin/matchmaker/__main__.py:27]: I'm pretty sure it's stderr

File: prologin/matchmaker/matchmaker.py [hunk 15]
```diff
@@ -0,0 +1,397 @@
...
+        """Runs the service."""
+        await self.setup()
+        await self.bootstrap()
+
+        # Start service components
+        await asyncio.wait(
+            [self.new_champions_watcher_loop(), self.create_matches_loop()]
+        )
+
+    def get_champions_query(self):
```
Comment (Zopieux) [inline prologin/matchmaker/matchmaker.py:154]: Nit: I'd move that under the setup & bootstrap.

File: prologin/matchmaker/matchmaker.py [hunk 16]
```diff
@@ -0,0 +1,397 @@
...
+            username=self.config["matchmaker"]["author_username"]
+        )
+        logging.info(
+            "MatchMaker will use %s as author to schedule new matches",
+            self.author,
+        )
+
+        self.tournament, _ = Tournament.objects.get_or_create(
+            name=self.config["matchmaker"]["tournament_name"],
+            defaults={"author": self.author},
```
Comment (Zopieux) [inline prologin/matchmaker/matchmaker.py:190]: What happens if the existing tournament author is not the one configured (`self.author`)?

File: prologin/matchmaker/matchmaker.py [hunk 17]
```diff
@@ -0,0 +1,397 @@
...
+        """Process backlog and set the service in a known state."""
+        # Get latest champions
+        latest_champions_query = self.get_champions_query()
+        latest_champions = set(latest_champions_query)
+        logging.info(
+            "bootstrap: managing %d champions in the tournament",
+            len(latest_champions),
+        )
+
+        # Remove champions that were the tournament but are not anymore
```
Comment (Zopieux) [inline prologin/matchmaker/matchmaker.py:210]: Missing a word

File: prologin/matchmaker/matchmaker.py [hunk 18]
```diff
@@ -0,0 +1,397 @@
...
+        match = MatchItem(champions=champions)
+        for r in range(self.config["matchmaker"].get("match_repeat", 1)):
+            if django_settings.STECHEC_USE_MAPS:
+                for map in self.maps:
+                    yield dataclasses.replace(match, map=map)
+            else:
+                yield match
+
+    def make_matches(self, matches) -> None:
+        """Prepare matches for scheduling."""
```
Comment (Zopieux) [inline prologin/matchmaker/matchmaker.py:276]: Prepares

File: prologin/matchmaker/matchmaker.py [hunk 19]
```diff
@@ -0,0 +1,397 @@
...
+        # Keep track of matches for each champion
+        for match in matches:
+            for champion in match.champions:
+                self.matches_with_champion[champion].add(match)
+        self.next_matches.extend(matches)
+
+    async def cancel_matches(self, old_champions: Iterable[Champion]) -> None:
+        """Cancel matches with old champions."""
```
Comment (Zopieux) [inline prologin/matchmaker/matchmaker.py:284]: Cancels

File: prologin/matchmaker/matchmaker.py [hunk 20]
```diff
@@ -0,0 +1,397 @@
...
+        logging.info(
+            'cancelling matches with old champions: %s', old_champions
+        )
+
+        # Cancel prepared matches
+        for old_champion in old_champions:
+            if old_champion not in self.matches_with_champion:
+                continue
+            for match in self.matches_with_champion.pop(old_champion):
```
Comment (Zopieux) [inline prologin/matchmaker/matchmaker.py:293]: `.pop(old_champion, [])` and drop that if :-)

File: prologin/matchmaker/tests.py [hunk 21]
```diff
@@ -0,0 +1,214 @@
...
+    MatchPriority,
+    Tournament,
+)
+
+from .matchmaker import MatchItem, MatchMaker
+
+
+class MatchMakerTest(TestCase):
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
```
Comment (Zopieux) [inline prologin/matchmaker/tests.py:23]: Not needed

File: prologin/matchmaker/tests.py [hunk 22]
```diff
@@ -0,0 +1,214 @@
...
+            len(
+                list(
+                    itertools.permutations(
+                        self._users, r=settings.STECHEC_NPLAYERS
+                    )
+                )
+            )
+            * self._mm_config['matchmaker']['match_repeat']
+            * len(self._maps)
+        )
```
Comment (Zopieux) [inline prologin/matchmaker/tests.py:36]: https://docs.python.org/3/library/math.html#math.perm

```python
>>> len(list(itertools.permutations("abcde", r=3)))
60
>>> math.perm(5, 3)
60
```

**Review by Zopieux — APPROVED** [review]
Just a few nits.

That looks so great! Thanks for adding tests.

**Comment by juli0z** [comment]
Migrated to https://gitlab.com/prologin/tech/infra/sadm/-/merge_requests/182
